### PR TITLE
Core Services Move to Mono Repo

### PR DESCRIPTION
**Pull request by jpwhitemn**
updated all core go file packages based on Mono repo, updated makefile for core services, added cmd folders for core services, moved Fede C fixes found in test repo

### CODE REVIEW

File: core/metadata/metadata.go [hunk 1]
```diff
@@ -1,55 +1,44 @@
...
- * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
- * or implied. See the License for the specific language governing permissions and limitations under
- * the License.
- *
- * @microservice: core-metadata-go service
- * @author: Spencer Bull & Ryan Comer, Dell
- * @version: 0.5.0
- *******************************************************************************/
-package main
+//
```
Comment (feclare) [inline core/metadata/metadata.go]: License header should not be changed in this case. At least it should not lost the dell copyright.

File: cmd/core-command/main.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright 2018 Dell Inc.
```
Comment (drasko) [inline cmd/core-command/main.go]: I propose [SPDX](https://spdx.org/) style headers, to keep consistency across the repo

File: cmd/core-data/main.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+var loggingClient logger.LoggingClient
+
+// Read the configuration file and update configuration struct
+func readConfigurationFile(path string) (*data.ConfigurationStruct, error) {
+	var configuration data.ConfigurationStruct
+	// Read the configuration file
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		fmt.Println("Error reading configuration file: " + err.Error())
```
Comment (drasko) [inline cmd/core-data/main.go]: Avoid printing on the screen - use logger

File: cmd/core-data/main.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		return nil, err
+	}
+
+	// Decode the configuration as JSON
+	err = json.Unmarshal(contents, &configuration)
+	if err != nil {
+		fmt.Println("Error reading configuration file: " + err.Error())
```
Comment (drasko) [inline cmd/core-data/main.go]: No stdout prints - log

File: cmd/core-data/main.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	return &configuration, nil
+}
+
+func main() {
+	start := time.Now()
+
+	// Load configuration data
+	configuration, err := readConfigurationFile(configFile)
+	if err != nil {
+		fmt.Println("Could not read config file: " + configFile)
```
Comment (drasko) [inline cmd/core-data/main.go]: Should be logged

File: cmd/core-metadata/main.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+func main() {
+	var loggingClient = logger.NewClient(metadata.METADATASERVICENAME, "")
+
+	// Load configuration data
+	configuration, err := readConfigurationFile(CONFIG)
+	if err != nil {
+		loggingClient.Error("Could not read configuration file(" + CONFIG + "): " + err.Error())
+		os.Exit(1)
```
Comment (drasko) [inline cmd/core-metadata/main.go]: Do we need to exit like this, or simple `return` from main will do?

File: core/command/const.go [hunk 7]
```diff
@@ -15,7 +15,7 @@
  * @author: Spencer Bull, Dell
  * @version: 0.5.0
  *******************************************************************************/
-package main
+package command
 
 // ConfigurationStruct : Struct used to pase the JSON configuration file
 type ConfigurationStruct struct {
```
Comment (drasko) [inline core/command/const.go]: No need to call it `ConfigurationStruct` - we see that it is `struct`. Go idiomatic way of naming variables suggest using very short and simple names.

**Comment by JPWKU** [comment]
recheck

**Comment by JPWKU** [comment]
recheck

File: core/command/command.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	"time"
+
+	consulclient "github.com/edgexfoundry/edgex-go/support/consul-client"
+	logger "github.com/edgexfoundry/edgex-go/support/logging-client"
+)
+
+var loggingClient logger.LoggingClient
+
+func heartbeat() {
+	for true {
```
[the remark this answers is not in the document]
Comment (drasko) [reply core/command/command.go]: Possibly better solution to heartbeat is using `time` Ticker functionality and listening on a channel: https://stackoverflow.com/questions/16466320/is-there-a-way-to-do-repetitive-tasks-at-intervals-in-golang

Something like this: https://github.com/mixer/redutil/blob/master/heartbeat/simple_heart.go#L91-L103

File: core/command/command.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		ServiceName:    configuration.ServiceName,
+		ServicePort:    configuration.ServerPort,
+		ServiceAddress: configuration.ServiceAddress,
+		CheckAddress:   configuration.ConsulCheckAddress,
+		CheckInterval:  configuration.CheckInterval,
+		ConsulAddress:  configuration.ConsulHost,
+		ConsulPort:     configuration.ConsulPort,
+	})
+
+	if err == nil { // Update configuration data from Consul
```
Comment (drasko) [inline core/command/command.go]: Drop the unnecessary comment

File: core/data/init.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+var dbc clients.DBClient
+var loggingClient logger.LoggingClient
+var ep *messaging.EventPublisher
+var mdc metadataclients.DeviceClient
+var msc metadataclients.ServiceClient
+
+// Heartbeat for the data microservice - send a message to logging service
+func heartbeat() {
+	// Loop forever
+	for true {
```
Comment (drasko) [inline core/data/init.go]: See comment above for Go idiomatic infinite loops: https://tour.golang.org/flowcontrol/4

File: core/data/init.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		ServiceName:    configuration.Servicename,
+		ServicePort:    configuration.Serverport,
+		ServiceAddress: configuration.Serviceaddress,
+		CheckAddress:   configuration.Consulcheckaddress,
+		CheckInterval:  configuration.Checkinterval,
+		ConsulAddress:  configuration.Consulhost,
+		ConsulPort:     configuration.Consulport,
+	})
+
+	if err != nil {
```
Comment (drasko) [inline core/data/init.go]: Do not use `if - else`. If `err` then log the err and return it (I see that your function should return error)

File: core/metadata/metadata.go [hunk 12]
```diff
@@ -1,55 +1,44 @@
...
-	start := time.Now()
-
-	// Load configuration data
-	err := readConfigurationFile(CONFIG)
-	if err != nil {
-		fmt.Printf("Could not read configuration file(%s): %#v\n", CONFIG, err)
-		os.Exit(1)
+// Heartbeat for the metadata microservice - send a message to logging service
+func heartbeat() {
+	// Loop forever
```
Comment (drasko) [inline core/metadata/metadata.go]: Comment bloats the code. It's obvious what code does.

File: core/metadata/metadata.go [hunk 13]
```diff
@@ -1,55 +1,44 @@
...
+	for true {
```
Comment (drasko) [inline core/metadata/metadata.go]: see comment above about Go idiomatic infinite loops.

**Comment by drasko** [comment]
@tsconn23 do you really approve this code, even though I listed a lot of remarks and things that should be corrected?

You are maintainer, it's your call, but usually the code that goes in the repo, stays there :). So I prefer that we filter it out here more meticulously and  improve the code quality.

File: cmd/core-command/main.go (hunk 2 above)
Comment (jpwhitemn) [reply cmd/core-command/main.go]: We addressed the format of the header before with LF.  Keeping it the same for now per agreement of community

File: cmd/core-data/main.go (hunk 4 above)
Comment (jpwhitemn) [reply cmd/core-data/main.go]: fixed

File: core/command/const.go (hunk 7 above)
Comment (jpwhitemn) [reply core/command/const.go]: leaving for now as it is more cosmetic, but will address as I do more refactoring going forward.

File: cmd/core-command/main.go (hunk 2 above)
Comment (drasko) [reply cmd/core-command/main.go]: @jpwhitemn LF proposed SPDX (while I was proposing no comments at all). So we moved to SPDX with `export` services.

This is less important as a remark, I have no objection that comments are kept by a preference Just that I suggested code style aligning.

File: core/command/const.go (hunk 7 above)
Comment (drasko) [reply core/command/const.go]: Agreed.

**Comment by drasko** [comment]
@jpwhitemn as you will not address all the remarks for now, please let us know when your changes are done, and PR is ready.

**Comment by tsconn23** [comment]
@drasko Drasko, Jim asked me directly to approve it so I did

File: core/metadata/metadata.go (hunk 13 above)
Comment (jpwhitemn) [reply core/metadata/metadata.go]: Drasko - let us just get the code moved over from the multiple repos before we start trying to clean up the world under one PR!  We agreed we just wanted to get mono repo up with existing code.  Let's get everything working before we burden the PRs with more requests.

**Comment by JPWKU** [comment]
recheck

**Comment by drasko** [comment]
@tsconn23 np, just wanted to check with you guys should I go further with the review. In any case - no functional issues found from my side, just these details, so we can proceed we ca proceed with the merge.

File: core/clients/coredataclients/coredata-clients.go [hunk 14]
```diff
@@ -72,6 +77,19 @@ func decodeValueDescriptorSlice(resp *http.Response) ([]models.ValueDescriptor,
...
 }
 
+// Helper method to decode a device and return the device
+func (v *ValueDescriptorClient) decodeValueDescriptor(resp *http.Response) (models.ValueDescriptor, error) {
+	dec := json.NewDecoder(resp.Body)
+	vdr := models.ValueDescriptor{}
+
+	err := dec.Decode(&vdr)
+	if err != nil {
+		fmt.Println(err)
```
Comment (drasko) [inline core/clients/coredataclients/coredata-clients.go]: Printing on the stdout should be avoided. Use logs instead.

Comment (drasko) [reply core/clients/coredataclients/coredata-clients.go]: You can also retun immediately here, with `return nil, err` - no need returning a struct below, when we have an error.

File: core/clients/coredataclients/coredata-clients.go [hunk 15]
```diff
@@ -103,5 +121,346 @@ func (v *ValueDescriptorClient) ValueDescriptors() ([]models.ValueDescriptor, er
...
 
-	return decodeValueDescriptorSlice(resp)
+	return v.decodeValueDescriptorSlice(resp)
+}
+
+// Get the value descriptor by id
+func (v *ValueDescriptorClient) ValueDescriptor(id string) (models.ValueDescriptor, error) {
+	req, err := http.NewRequest("GET", v.url+"/"+id, nil)
+	if err != nil {
+		fmt.Println(err)
```
Comment (drasko) [inline core/clients/coredataclients/coredata-clients.go]: Should not print on stdout. Use logs.

File: core/clients/coredataclients/coredata-clients.go [hunk 16]
```diff
@@ -103,5 +121,346 @@ func (v *ValueDescriptorClient) ValueDescriptors() ([]models.ValueDescriptor, er
...
+		return models.ValueDescriptor{}, err
```
Comment (drasko) [inline core/clients/coredataclients/coredata-clients.go]: No need to return empty struct on error. Return `nil`.

**Comment by feclare** [comment]
recheck

**Pull request closed without merging**